File: server/src/controllers/auth.controller.ts
import { Request, Response, NextFunction } from 'express';
import { AuthService } from '../services/auth.service.js';
import {
  registerSchema,
  loginSchema,
  loginOtpSchema,
  sendOtpSchema,
  verifyOtpSchema,
  googleAuthSchema,
} from '../validators/index.js';
import { AuthenticatedRequest } from '../middleware/auth.middleware.js';
import { config } from '../config/env.js';

export class AuthController {
  /**
   * Request OTP code to be sent to user's Gmail (for registration or login)
   */
  static async sendOtp(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const { email, purpose, name } = sendOtpSchema.parse(req.body);
      const result = await AuthService.sendOtp(email, purpose, name);

      res.status(200).json({
        success: true,
        message: `6-digit verification code sent to ${result.email}`,
        data: { email: result.email },
      });
    } catch (error) {
      next(error);
    }
  }

  /**
   * Verify an OTP code standalone
   */
  static async verifyOtp(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const { email, otp, purpose } = verifyOtpSchema.parse(req.body);
      await AuthService.verifyOtp(email, otp, purpose);

      res.status(200).json({
        success: true,
        message: 'OTP verified successfully',
      });
    } catch (error) {
      next(error);
    }
  }

  /**
   * User Registration (Sign Up) with OTP verification
   */
  static async register(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const validated = registerSchema.parse(req.body);
      const { user, token } = await AuthService.register(validated);

      res.cookie('token', token, {
        httpOnly: true,
        secure: config.env === 'production',
        sameSite: 'lax',
        maxAge: 7 * 24 * 60 * 60 * 1000, // 7 days
      });

      res.status(201).json({
        success: true,
        message: 'Account registered successfully 💖',
        data: { user, token },
      });
    } catch (error) {
      next(error);
    }
  }

  /**
   * User Login Step 1: Validates credentials and sends 6-digit OTP to user's Gmail
   */
  static async initiateLogin(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const { identifier, password } = loginSchema.parse(req.body);
      const result = await AuthService.initiateLogin(identifier, password);

      res.status(200).json({
        success: true,
        message: `Verification code sent to your Gmail (${result.email})`,
        data: { email: result.email, name: result.name, requireOtp: true },
      });
    } catch (error) {
      next(error);
    }
  }

  /**
   * User Login Step 2: Complete Login after verifying 6-digit OTP
   */
  static async completeLoginWithOtp(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const { email, otp } = loginOtpSchema.parse(req.body);
      const { user, token } = await AuthService.completeLoginWithOtp(email, otp);

      res.cookie('token', token, {
        httpOnly: true,
        secure: config.env === 'production',
        sameSite: 'lax',
        maxAge: 7 * 24 * 60 * 60 * 1000,
      });

      res.status(200).json({
        success: true,
        message: 'Signed in successfully 💖',
        data: { user, token },
      });
    } catch (error) {
      next(error);
    }
  }

  /**
   * Direct password login (supports fallback)
   */
  static async login(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const { identifier, password } = loginSchema.parse(req.body);
      const { user, token } = await AuthService.login(identifier, password);

      res.cookie('token', token, {
        httpOnly: true,
        secure: config.env === 'production',
        sameSite: 'lax',
        maxAge: 7 * 24 * 60 * 60 * 1000,
      });

      res.status(200).json({
        success: true,
        message: 'Signed in successfully',
        data: { user, token },
      });
    } catch (error) {
      next(error);
    }
  }

  /**
   * Google ID token authentication
   */
  static async googleLogin(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const { idToken } = googleAuthSchema.parse(req.body);
      const { user, token } = await AuthService.verifyGoogleIdToken(idToken);

      res.cookie('token', token, {
        httpOnly: true,
        secure: config.env === 'production',
        sameSite: 'lax',
        maxAge: 7 * 24 * 60 * 60 * 1000,
      });

      res.status(200).json({
        success: true,
        message: 'Authenticated successfully with Google',
        data: { user, token },
      });
    } catch (error) {
      next(error);
    }
  }

  /**
   * Get current authenticated user session
   */
  static async getMe(req: AuthenticatedRequest, res: Response, next: NextFunction): Promise<void> {
    try {
      res.status(200).json({
        success: true,
        data: { user: req.user },
      });
    } catch (error) {
      next(error);
    }
  }

  /**
   * Logout user and clear session cookie
   */
  static async logout(_req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      res.clearCookie('token');
      res.status(200).json({
        success: true,
        message: 'Logged out successfully',
      });
    } catch (error) {
      next(error);
    }
  }
}

File: server/src/main/java/com/mchat/controller/AuthController.java
package com.mchat.controller;

import com.mchat.dto.ApiResponse;
import com.mchat.dto.AuthDTO.*;
import com.mchat.security.UserPrincipal;
import com.mchat.service.AuthService;
import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/auth")
@RequiredArgsConstructor
public class AuthController {

    private final AuthService authService;

    @PostMapping("/send-otp")
    public ResponseEntity<?> sendOtp(@Valid @RequestBody SendOtpRequest req) {
        String email = authService.sendOtp(req.getEmail(), req.getPurpose(), req.getName());
        return ResponseEntity.ok(ApiResponse.ok("6-digit verification code sent to " + email, Map.of("email", email)));
    }

    @PostMapping("/verify-otp")
    public ResponseEntity<?> verifyOtp(@Valid @RequestBody VerifyOtpRequest req) {
        authService.verifyOtp(req.getEmail(), req.getOtp(), req.getPurpose());
        return ResponseEntity.ok(ApiResponse.ok("OTP verified successfully", null));
    }

    @PostMapping("/register")
    public ResponseEntity<?> register(@Valid @RequestBody RegisterRequest req, HttpServletResponse response) {
        AuthResponse auth = authService.register(req);
        attachTokenCookie(response, auth.getToken());
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.ok("Account registered successfully 💖", auth));
    }

    @PostMapping("/login-init")
    public ResponseEntity<?> initiateLogin(@Valid @RequestBody LoginInitRequest req) {
        String email = authService.initiateLogin(req.getIdentifier(), req.getPassword());
        return ResponseEntity.ok(ApiResponse.ok(
                "Verification code sent to your Gmail (" + email + ")",
                Map.of("email", email, "requireOtp", true)
        ));
    }

    @PostMapping("/login-otp")
    public ResponseEntity<?> completeLoginWithOtp(@Valid @RequestBody LoginOtpRequest req, HttpServletResponse response) {
        AuthResponse auth = authService.completeLoginWithOtp(req.getEmail(), req.getOtp());
        attachTokenCookie(response, auth.getToken());
        return ResponseEntity.ok(ApiResponse.ok("Signed in successfully 💖", auth));
    }

    @PostMapping("/login")
    public ResponseEntity<?> login(@Valid @RequestBody LoginRequest req, HttpServletResponse response) {
        AuthResponse auth = authService.login(req.getIdentifier(), req.getPassword());
        attachTokenCookie(response, auth.getToken());
        return ResponseEntity.ok(ApiResponse.ok("Signed in successfully", auth));
    }

    @PostMapping("/google")
    public ResponseEntity<?> googleLogin(@Valid @RequestBody GoogleLoginRequest req, HttpServletResponse response) {
        AuthResponse auth = authService.googleLogin(req.getIdToken());
        attachTokenCookie(response, auth.getToken());
        return ResponseEntity.ok(ApiResponse.ok("Authenticated successfully with Google", auth));
    }

    @PostMapping("/dev-login")
    public ResponseEntity<?> devLogin(@RequestBody(required = false) Map<String, String> body, HttpServletResponse response) {
        String persona = body != null ? body.get("persona") : "Alice";
        AuthResponse auth = authService.devLogin(persona);
        attachTokenCookie(response, auth.getToken());
        return ResponseEntity.ok(ApiResponse.ok("Logged in as dev persona " + persona, auth));
    }

    @GetMapping("/me")
    public ResponseEntity<?> getMe(@AuthenticationPrincipal UserPrincipal principal) {
        if (principal == null) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(ApiResponse.error("Not authenticated"));
        }
        return ResponseEntity.ok(ApiResponse.ok(Map.of("user", principal.getUser())));
    }

    @PostMapping("/logout")
    public ResponseEntity<?> logout(HttpServletResponse response) {
        Cookie cookie = new Cookie("token", "");
        cookie.setPath("/");
        cookie.setHttpOnly(true);
        cookie.setMaxAge(0);
        response.addCookie(cookie);
        return ResponseEntity.ok(ApiResponse.ok("Logged out successfully", null));
    }

    private void attachTokenCookie(HttpServletResponse response, String token) {
        Cookie cookie = new Cookie("token", token);
        cookie.setPath("/");
        cookie.setHttpOnly(true);
        cookie.setMaxAge(7 * 24 * 60 * 60); // 7 days
        response.addCookie(cookie);
    }
}
